### DISCUSSION TITLE
Use Espresso instead of the Java Virtual Machine. How to parse native methods?

### DISCUSSION THREAD

**Discussion by tcmot** [oracle/graal#8283] [General] [answered]
```
 java.lang.System

@HotSpotIntrinsicCandidate
 public static native long currentTimeMillis();
```

When using the above code, theJava Virtual Machine will parse it in C code.
How does a graalvm Espresso do it?

Thanks.

**Comment by gilles-duboscq** [comment, marked as answer]
For this specific one, there is an espresso-specific "substitution".
You can see it in [Target_java_lang_System.java](https://github.com/oracle/graal/blob/master/espresso/src/com.oracle.truffle.espresso/src/com/oracle/truffle/espresso/substitutions/Target_java_lang_System.java#L513-L516). When a substituted method is called, espresso will run the code of the substitution in the host jvm.

In general though, for native methods, espresso supports JNI and will load and link to native libraries like any other jvm implementation. Espresso can use Truffle NFI or Sulong for loading the native libraries. When using NFI, it'll call the native code in the usual way. If you have a version of the library with bitcode, it can be executed by Sulong.

You can check out the [NativeAccess](https://github.com/oracle/graal/blob/master/espresso/src/com.oracle.truffle.espresso/src/com/oracle/truffle/espresso/ffi/NativeAccess.java) interface an its implementations.

**Reply by tcmot** [reply]
Thank you for your reply.
Anyway, the virtual machine must load the native library(operating system Thread，File，etc) through JNI. 

How does NativeImage parse native methods during compilation?

**Reply by tcmot** [reply]
NativeImage tool compiles Java source code into x86 code and other?

**Reply by gilles-duboscq** [reply]
Native image also implements JNI. It compiles java bytecode to native code. If java code calls a native method, the native code from the corresponding JNI library will be called (probably through a small stub that deals with JNI details like allocating handles, adapting the calling convention etc.).

Native image also has a substitution mechanism which means some things might not use the standard C implementation found in OpenJDK. Native image also has "system java" which allows directly interacting with native code and data structures without going through JNI.

**Reply by tcmot** [reply]
Thank you for your reply.
